HMS-5287: Add playwright and github action to frontend (#416)
content-services/content-sources-frontend: README.md

To setup the hosts file run following command:

```bash
yarn patch:hosts
```

> Alternativly, simply add these lines to your /etc/hosts:

Reviewer: ```suggestion
Alternatively, add these lines to your /etc/hosts file:
```
In technical writing, don't use words like simple, easy or quick.  It is subjective.

## Getting started - Installation

1. Make sure [nvm](https://github.com/nvm-sh/nvm) is installed

2. To ensure you have the correct node version installed do: `nvm use`.
   Yarn WILL prevent you from progressing if you have not updated your node version to match the one in the .nvmrc file.

3. `yarn install`

4. `yarn build` (only required when setting up for the first time)

## Getting started - Running the app

Keep in mind that you have to be connected to the VPN for this to work, even in the offices.

1. `yarn start` to choose whether to run against stage or prod environments. <br/>
   OR <br/>
   `yarn start:stage` to run against stage environment. <br/>
   OR <br/>
   `yarn start:prod` to run against prod environment. <br/>
   OR <br/>
   `yarn local` to run against a local backend running on port 8000.<br/>

2. With a browser, open the URL listed in the terminal output, https://stage.foo.redhat.com:1337/insights/content for example.

### Unit Testing

`yarn verify` will run `yarn build` `yarn lint` (eslint), `yarn format:check` Prettier formatting check and `yarn test` (Jest unit tests)

One can also: `yarn test` to run the unit tests directly.

## Testing with Playwright

> 1. Ensure to `yarn playwright install --with-deps` this installs the required browsers (or vnc browsers) to view/run playwright front-end tests.

Reviewer 2: as mentioned on the other PR  --with-deps won't work on fedora, but it works fine without, should we just drop it?

Author: Likely should mention this step isn't needed in "fedora" specifically.

2. Run the backend locally, steps to do this can be found in the [backend repository](https://github.com/content-services/content-sources-backend).

   Ensure that the backend is running prior to the following steps.

3. `yarn local` will start up the front-end repository. If you do `yarn start` and choose stage, your tests will attempt to run against the stage ENV.

> 4. `yarn playwright test` will spin-off all of the playwright test suite. `yarn playwright test --headed` will run the suite in a vnc-like browser so you can follow what it is doing.

Reviewer 2: should probably add a step to setup the .env file, making sure to specify 

```
BASE_URL=https://stage.foo.redhat.com:1337
```

Author: There isn't one? 🤔

It is recommended to test using vs-code and the [Playwright Test module for VSCode](https://marketplace.visualstudio.com/items?itemName=ms-playwright.playwright).

## PR checks and linking front-end/backend-end PRs for testing

> The CICD pipeline for playwright (both front-end and backend) will check for presence in the description of the front-end prs, for the following formatted text:

Reviewer: ```suggestion
The CICD pipeline for playwright (both front-end and backend) will check in the description of the front-end PRs for the following formatted text:
```

`#testwith https://github.com/content-services/content-sources-backend/pull/<PR NUMBER>`

> Note: the space between `#testwith https`.

Reviewer: ```suggestion
Note the space in `#testwith https`.
```

> If a backend PR is linked, the front-end/back-end PR's in question will both use the corresponding linked branch for their Playwright tests in the PR check.

Reviewer: ```suggestion
If a backend PR is linked, the front-end and back-end PR's in question will both use the corresponding linked branch for their Playwright tests in the PR check.
```

::1 ci.foo.redhat.com
```

## Getting started - Installation

1. Make sure [nvm](https://github.com/nvm-sh/nvm) is installed

> 2. First time running the app do: `nvm use` to ensure you have the correct node version installed.

Reviewer: ```suggestion
2. First time running the app do: `nvm use` to ensure you have the correct node version installed.
```
If this is important then no need to say "First time running the app do:". Rather tell the user what version they need and say sonmehting short and to the point, e.g. "Ensure node version X is in use"

   If you do not, follow the instructions nvm gives you to install the appropriate version.
   Yarn WILL prevent you from progressing if you have not updated your node version.

3. `yarn install`

4. `yarn build` (only required when setting up for the first time)

## Getting started - Running the app

Keep in mind that you have to be connected to the VPN for this to work, even in the offices.

1. `yarn start` to choose whether to run against stage or prod environments. <br/>
   OR <br/>
   `yarn start:stage` to run against stage environment. <br/>
   OR <br/>
   `yarn start:prod` to run against prod environment. <br/>
   OR <br/>

>    `yarn local` to run against our local backend running on port 8000.<br/>

Reviewer: "our local backend" is a bit vague. I see later it says to run it locally.

2. With a browser, open the URL listed in the terminal output, https://stage.foo.redhat.com:1337/insights/content for example.

### Unit Testing

`yarn verify` will run `yarn build` `yarn lint` (eslint), `yarn format:check` Prettier formatting check and `yarn test` (Jest unit tests)

> Alternatively one can run simply: `yarn test` to run the unit tests.

Reviewer: ```suggestion
Run `yarn test` to run the unit tests.
```"simply" is marketing speak. In technical writing, never make judgements about how hard or easy something is.

## Testing with Playwright

1. Ensure to `yarn playwright install --with-deps` this installs the required browsers (or vnc browsers) to view/run playwright front-end tests.

2. Run the backend locally, steps to do this can be found in the [backend repository](https://github.com/content-services/content-sources-backend).
   Eensure that the backend is running prior to the following steps.

3. `yarn local` will start up the front-end repository. If you do `yarn start` and choose stage, your tests will attempt to run against the stage ENV.

4. `yarn playwright test` will spin-off all of the playwright test suite. `yarn playwright test --headed` will run the suite in a vnc-like browser so you can follow what it is doing.

> I highly recommend testing using vs-code and the [Playwright Test module for VSCode](https://marketplace.visualstudio.com/items?itemName=ms-playwright.playwright).

Reviewer 3: `I`? :)
